### PR TITLE
RulesCrudApi

### PR DESCRIPTION
**Pull request by rscronce**

### CODE REVIEW

**Review by sarabjotanand — APPROVED** [review]

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+            },
+            "previewLimit": 50,
+            "server": "ralph-database-1.c1idjdax3vgr.us-east-1.rds.amazonaws.com",
+            "port": 3306,
+            "driver": "MySQL",
+            "database": "ralph-database-1",
+            "username": "admin",
+            "password": "agile2020",
+            "name": "Ralph-Database-1"
+        },
```
Comment (jamesfulford) [inline .vscode/settings.json:15]: I doubt you wanted to commit this

File: backend/src/views.py [hunk 2]
```diff
@@ -5,6 +5,55 @@
...
+
+@csrf_exempt
+def delete_rule(request, rule_id):
+    # user = request.user.id
+    try:
+        rule = Rule.objects.get(id=rule_id)
+        rule.delete()
+        return JsonResponse({'message': 'Rule was successfully deleted'}, status=status.HTTP_204_NO_CONTENT)
+    except ObjectDoesNotExist as e:
+        return JsonResponse({'error': str(e)}, safe=False, status=status.HTTP_404_NOT_FOUND)
```
Comment (jamesfulford) [inline backend/src/views.py:53]:
```suggestion
    except ObjectDoesNotExist as e:
        return JsonResponse({'error': 'Rule not found' }, status=status.HTTP_404_NOT_FOUND)       
```
1. safe should not be False.
2. We should not sent stringified exceptions in a response, except in very careful circumstances. See other comment.

File: backend/src/views.py [hunk 3]
```diff
@@ -5,6 +5,55 @@
...
+        return JsonResponse(rule_serializer.data, status=status.HTTP_201_CREATED) 
+    return JsonResponse(rule_serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+@csrf_exempt
+def delete_rule(request, rule_id):
+    # user = request.user.id
+    try:
+        rule = Rule.objects.get(id=rule_id)
+        rule.delete()
+        return JsonResponse({'message': 'Rule was successfully deleted'}, status=status.HTTP_204_NO_CONTENT)
```
Comment (jamesfulford) [inline backend/src/views.py:51]:
```suggestion
        return HttpResponse(status=status.HTTP_204_NO_CONTENT)
```

Add this import:
```
from django.http import HttpResponse
```

If the response is "204 No Content", we should not return a message with content. We should return an empty response (not an empty object response; an empty response)

**Review by jamesfulford — CHANGES_REQUESTED** [review]

**Comment by jamesfulford** [comment]
I didn't realize how much I've learned over the years. REST is an important requirement that I didn't voice, and they don't really teach it in school. API security is accumulated knowledge, also not taught. Not committing credentials means rest at night. Making dev envs that work out of the box pays off in the short run.

If it takes longer to do these stories, that's fine, as the PO I can adjust. (I'll definitely be chunking next sprint's stories down much further.) But I'd much rather pass on some lessons here in a school project than clean it up after the fact at work.

File: backend/src/views.py (hunk 2 above)
Comment (jamesfulford) [reply backend/src/views.py:53]: "other comment": https://github.com/MoneyWise2020/MoneyWise2020/pull/31#discussion_r521113563

File: backend/MoneyWise2020/settings.py [hunk 4]
```diff
@@ -22,7 +22,8 @@
 
 SECRET_KEY = os.environ['DJANGO_SECRET_KEY']
 
-DEBUG = False
+DEBUG = True
```
[the remark this answers is not in the document]
Comment (alejandrojsuazohvd) [reply backend/MoneyWise2020/settings.py]: I'm back. Changed this back to False.

File: backend/MoneyWise2020/settings.py [hunk 5]
```diff
@@ -134,3 +138,8 @@
 
 STATIC_ROOT = os.path.join(BASE_DIR, 'static')
 STATIC_URL = '/static/'
+
+CORS_ORIGIN_ALLOW_ALL = False
+CORS_ORIGIN_WHITELIST = (
+    'http://localhost:8001',
```
[the remark this answers is not in the document]
Comment (alejandrojsuazohvd) [reply backend/MoneyWise2020/settings.py]: Changed this to use 8080

File: backend/src/views.py [hunk 6]
```diff
@@ -5,6 +5,55 @@
...
+        return JsonResponse(rule_serializer.data)
+    return JsonResponse(rule_serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+    
+@csrf_exempt
+def get_rule_list(request):
+    rules = Rule.objects.all()
+    rule_serializer = RuleSerializer(rules, many=True)
+    return JsonResponse(rule_serializer.data, safe=False)
+
+def get_rule(request, rule_id):
```
[the remark this answers is not in the document]
Comment (alejandrojsuazohvd) [reply backend/src/views.py:33]: Added the annotation just in case.

File: backend/src/views.py [hunk 7]
```diff
@@ -5,6 +5,55 @@
...
+def delete_rule(request, rule_id):
+    # user = request.user.id
+    try:
+        rule = Rule.objects.get(id=rule_id)
+        rule.delete()
+        return JsonResponse({'message': 'Rule was successfully deleted'}, status=status.HTTP_204_NO_CONTENT)
+    except ObjectDoesNotExist as e:
+        return JsonResponse({'error': str(e)}, safe=False, status=status.HTTP_404_NOT_FOUND)       
+    except Exception as e:
+        return JsonResponse({'error': str(e)}, safe=False, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
```
[the remark this answers is not in the document]
Comment (alejandrojsuazohvd) [reply backend/src/views.py]: I updated this to return a generic error message to the client and also added logging to the views to ensure we can at least capture the error within our internal logs.

File: backend/src/views.py [hunk 8]
```diff
@@ -5,6 +5,55 @@
...
+from rest_framework import status
+from .serializers import RuleSerializer
+from rest_framework.parsers import JSONParser 
+
+@csrf_exempt
+def update_rule(request, rule_id):
+    rule = Rule.objects.get(id=rule_id)
+    rule_data = JSONParser().parse(request)
+    rule_serializer = RuleSerializer(rule, data=rule_data)
+    if rule_serializer.is_valid():
```
[the remark this answers is not in the document]
Comment (rscronce) [reply backend/src/views.py:21]: According to the documentation, "is_valid" should be called prior to persisting the data. It validates all required fields are provided and have correct types.

File: backend/src/urls.py [hunk 9]
```diff
@@ -4,5 +4,10 @@
 urlpatterns = [
     path('api/ping', views.hello_world),
     path('api/transactions', views.process_transactions),
-#    path('createrule', views.createrule),
+    path('api/delete_rule/<str:rule_id>', views.delete_rule),
+    path('api/create_rule', views.create_rule),
+    path('api/get_rule/<str:rule_id>', views.get_rule),    
+    path('api/get_rule_list', views.get_rule_list),    
+    path('api/update_rule/<str:rule_id>', views.update_rule),
```
[the remark this answers is not in the document]
Comment (rscronce) [reply backend/src/urls.py:11]: I think you're looking for like the below in VIEWS.PY.  Is this a correct understanding?

**def rules_handler(request):**
    if request.method == 'GET': 
        get_rule_list(request)
    if request.method == "POST":
        create_rule(request)

**def rules_by_id_handler(request, rule_id):**    
    if request.method == 'GET': 
        get_rule(request, rule_id)
    elif request.method == 'DELETE':
        delete_rule(request, rule_id)
    elif request.method == 'PUT':
        update_rule(request, rule_id)def rules_handler(request):
    if request.method == 'GET': 
        get_rule_list(request)
    if request.method == "POST":
        create_rule(request)

def rules_by_id_handler(request, rule_id):    
    if request.method == 'GET': 
        get_rule(request, rule_id)
    elif request.method == 'DELETE':
        delete_rule(request, rule_id)
    elif request.method == 'PUT':
        update_rule(request, rule_id)

File: .vscode/settings.json (hunk 1 above)
Comment (rscronce) [reply .vscode/settings.json:15]: I believe the latest push addressed this concern.

File: .vscode/settings.json [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+            },
+            "previewLimit": 50,
+            "server": "moneywise-database-1.c1idjdax3vgr.us-east-1.rds.amazonaws.com",
+            "port": 3306,
+            "driver": "MySQL",
+            "name": "MoneyWise",
+            "username": "admin",
+            "password": "agile2020",
+            "database": "MoneyWise"
+        }
```
[the remark this answers is not in the document]
Comment (rscronce) [reply .vscode/settings.json:28]: I believe the latest push addressed this concern.

File: backend/src/urls.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rscronce) [reply backend/src/urls.py:11]: I believe the latest push addressed this concern.

File: backend/src/views.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jamesfulford) [reply backend/src/views.py:38]: OK, just type checks. Got it.

**Review by jamesfulford — APPROVED** [review]

**Pull request merged**